### ISSUE TITLE
UnsatisfiedLinkError on Alpine linux

### ISSUE THREAD

**Issue by htmldoug** [luben/zstd-jni#43]
Running zstd-jni 1.3.0-1 in the [openjdk:8-jre-alpine](https://hub.docker.com/_/openjdk/) container on an ubuntu host produces a `java.lang.UnsatisfiedLinkError`.

```
java.lang.UnsatisfiedLinkError: no zstd-jni in java.library.path
Unsupported OS/arch, cannot find /linux/amd64/libzstd-jni.so or load zstd-jni from system libraries. Please try building from source the jar or providing libzstd-jni in you system.
        at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
        at java.lang.Runtime.loadLibrary0(Runtime.java:870)
        at java.lang.System.loadLibrary(System.java:1122)
        at com.github.luben.zstd.util.Native.load(Native.java:101)
        at com.github.luben.zstd.ZstdOutputStream.<clinit>(ZstdOutputStream.java:17)
```

The debian based versions, `openjdk:slim` and `openjdk:<version>` both work fine.

From the description of `openjdk:alpine`:
> The main caveat to note is that it does use musl libc instead of glibc and friends, so certain software might run into issues depending on the depth of their libc requirements.

Looks like zstd-jni falls into this category.

Native.java selects `/linux/amd64/libzstd-jni.so`. I'm not sure what'd you need to do to determine that you should swap in a different module, but I'd be happy to gather more info about the environment from the jvm's perspective if it'd help.

An acceptable resolution to this issue might just be to update the readme to say that alpine is unsupported to save the next person some debugging time. Happy to submit a PR for that too, if you'd like me to.

**Comment by luben** [comment]
Yes, it has a few dependencies into libc. What happens if you build it on top alpine linux with musl libc?

**Comment by luben** [comment]
BTW, here is what's linked from glibc:

```
~/zstd-jni/target/classes/linux/amd64(master) » strings libzstd-jni.so | grep GLIBC                                                                                                                    luben@zenbook
GLIBC_2.3.4
GLIBC_2.14
GLIBC_2.4
GLIBC_2.2.5
free@@GLIBC_2.2.5
qsort@@GLIBC_2.2.5
clock@@GLIBC_2.2.5
__stack_chk_fail@@GLIBC_2.4
__assert_fail@@GLIBC_2.2.5
memset@@GLIBC_2.2.5
memcmp@@GLIBC_2.2.5
calloc@@GLIBC_2.2.5
memcpy@@GLIBC_2.14
malloc@@GLIBC_2.2.5
fflush@@GLIBC_2.2.5
memmove@@GLIBC_2.2.5
__fprintf_chk@@GLIBC_2.3.4
__cxa_finalize@@GLIBC_2.2.5
stderr@@GLIBC_2.2.5
```

**Comment by hvanhovell** [comment]
This can be caused by a `ClassLoader` issue. The current code may not use the correct `ClassLoader` and as a result it cannot find the resource. This can be fixed by changing https://github.com/luben/zstd-jni/blob/master/src/main/java/com/github/luben/zstd/util/Native.java#L65 to:
```java
// Note that resourceName must not be start with a '/'
InputStream is = Native.class.getClassLoader().getResourceAsStream(resourceName);
```

**Comment by hvanhovell** [comment]
Sorry for creating a fuss. The native libraries packaged in the zstd-jni library are compiled with glbc 2.4, which is a newer version than available on our build machines (CentOS 6.9).

**Comment by htmldoug** [comment]
Quick update:
I'm not able to reproduce this issue anymore over zstd-jni 1.3.0-1 and the latest openjdk:8-jre-alpine. I have no idea what changed. Hoping it's a change in 8-jre-alpine and not in my test methodology. I'll gain more confidence in the next few days as my switch back to alpine goes through QA.

I wrote a couple sbt tasks to run my zstd-jni scalatests under different docker containers. They pass. Sharing for the sake of sharing: https://gist.github.com/htmldoug/01e879a367a3b5a1d5e0155f45b23a7d

For the record, I tried building zstd-jni from source on alpine with `apk add alpine-sdk` and it built fine. I stopped short of testing it when I discovered that I couldn't repro the issue with the maven build.
